odoo/osv/expression.py

```diff
@@ -330,6 +339,374 @@ def distribute_not(domain):
+    ('&', [
+        ('partner_id', 'any', ('&', [('&', [('company_id', 'any', ('&', [('&', [('name', 'like', 'SGC')])]))])])),
+        ('|', [
+            ('partner_id', 'any', ('|', [
+                ('&', [('name', 'like', 'Jack')]),
+                ('&', [('name', 'like', "O'Neill")]),
+                ('&', [('function', '=', 'General')])
+            ]))
+        ])
+    ])
+    """
+
+    def _untreeify(domain_op, children):
+        domain = [domain_op] * (len(children) - (0 if domain_op == NOT_OPERATOR else 1))
+        for child in children:
+            if is_leaf(child):
+                fname, op, value = child
+                if op in ('any', 'not any'):
+                    domain.append((fname, op, _untreeify(*value)))
+                else:
+                    domain.append(child)
+            else:
+                domain.extend(_untreeify(*child))
+        return domain
+
+    new_domain = _untreeify(*merged_tree)
+    if _logger_any.isEnabledFor(logging.DEBUG):
+        _logger_any.debug('new_domain:\n%s\n', prettify_domain(new_domain))
+    """
+    ['&',
+        ('partner_id', 'any', [('company_id', 'any', [('name', 'like', 'SGC')])]),
+        ('partner_id', 'any', ['|', '|',
+            ('name', 'like', 'Jack'),
+            ('name', 'like', "O'Neill"),
+            ('function', '=', 'General')])]
+    """
+
+    if is_domain_stable:
+        return new_domain
+
+    # _combine_anys is sub-optimal for many2ones as it only combine anys
+    # for a same operator, multiple iterations are necessary to combine
+    # alternating AND and OR applied on compatible anys.
+    #
+    # Given the current algorithm, it is necessary to treeify/untreeify
+    # everytime. _treeify simplifies ['!', 'any'] => ['not any'] and
+    # _untreeify simplifies the many single-leaf-and-trees artifacts
+    # added by _combine_anys. Ideally, we should run multiple iterations
+    # of _combine_anys, RCO suggested using a real Tree object (instead
+    # of tuples) that could be smart and translate ['!', 'any'] and
+    # remove the artifacts added by _combine_anys on the fly.
+
+    guard = 9  # arbitrary
+    for _ in range(guard):
+        is_domain_stable = True
+
+        nested_tree = _treeify(new_domain)
+        if _logger_any.isEnabledFor(logging.DEBUG):
+            _logger_any.debug('nested_tree:\n%s', pprint.pformat(nested_tree))
+
+        dense_tree = _regroup(*nested_tree)
+        if _logger_any.isEnabledFor(logging.DEBUG):
+            _logger_any.debug('dense_tree:\n%s', pprint.pformat(dense_tree))
+
+        merged_tree = _combine_anys(*dense_tree, model)
+        if _logger_any.isEnabledFor(logging.DEBUG):
+            _logger_any.debug('merged_tree:\n%s', pprint.pformat(merged_tree))
+
+        new_domain = _untreeify(*merged_tree)
+        if _logger_any.isEnabledFor(logging.DEBUG):
+            _logger_any.debug('new_domain:\n%s\n', prettify_domain(new_domain))
```

Comment:
I have a problem with making these steps in a loop: 3/4 of it is pure overhead.
Conceptually, the algorithm should look like:
```
tree = domain2tree(domain)
repeat as many times as necessary:
    tree = combine_anys(tree)
return tree2domain(tree)
```